scipy/linalg/tests/test_matfuncs.py

```diff
@@ -437,69 +437,64 @@ def test_gh17918(self):
 
     def test_data_size_preservation_uint_in_float_out(self):
         M = np.zeros((10, 10), dtype=np.uint8)
-        # input bit size is 8, but minimum float bit size is 16
-        assert sqrtm(M).dtype == np.float16
+        assert sqrtm(M).dtype == np.float64
         M = np.zeros((10, 10), dtype=np.uint16)
-        assert sqrtm(M).dtype == np.float16
+        assert sqrtm(M).dtype == np.float64
         M = np.zeros((10, 10), dtype=np.uint32)
-        assert sqrtm(M).dtype == np.float32
+        assert sqrtm(M).dtype == np.float64
         M = np.zeros((10, 10), dtype=np.uint64)
         assert sqrtm(M).dtype == np.float64
 
     def test_data_size_preservation_int_in_float_out(self):
         M = np.zeros((10, 10), dtype=np.int8)
-        # input bit size is 8, but minimum float bit size is 16
-        assert sqrtm(M).dtype == np.float16
+        assert sqrtm(M).dtype == np.float64
         M = np.zeros((10, 10), dtype=np.int16)
-        assert sqrtm(M).dtype == np.float16
+        assert sqrtm(M).dtype == np.float64
         M = np.zeros((10, 10), dtype=np.int32)
-        assert sqrtm(M).dtype == np.float32
+        assert sqrtm(M).dtype == np.float64
         M = np.zeros((10, 10), dtype=np.int64)
         assert sqrtm(M).dtype == np.float64
 
     def test_data_size_preservation_int_in_comp_out(self):
         M = np.array([[2, 4], [0, -2]], dtype=np.int8)
-        # input bit size is 8, but minimum complex bit size is 64
-        assert sqrtm(M).dtype == np.complex64
+        assert sqrtm(M).dtype == np.complex128
         M = np.array([[2, 4], [0, -2]], dtype=np.int16)
-        # input bit size is 16, but minimum complex bit size is 64
-        assert sqrtm(M).dtype == np.complex64
+        assert sqrtm(M).dtype == np.complex128
         M = np.array([[2, 4], [0, -2]], dtype=np.int32)
-        assert sqrtm(M).dtype == np.complex64
+        assert sqrtm(M).dtype == np.complex128
         M = np.array([[2, 4], [0, -2]], dtype=np.int64)
         assert sqrtm(M).dtype == np.complex128
 
     def test_data_size_preservation_float_in_float_out(self):
         M = np.zeros((10, 10), dtype=np.float16)
-        assert sqrtm(M).dtype == np.float16
+        assert sqrtm(M).dtype == np.float32
         M = np.zeros((10, 10), dtype=np.float32)
         assert sqrtm(M).dtype == np.float32
         M = np.zeros((10, 10), dtype=np.float64)
         assert sqrtm(M).dtype == np.float64
         if hasattr(np, 'float128'):
             M = np.zeros((10, 10), dtype=np.float128)
-            assert sqrtm(M).dtype == np.float128
+            assert sqrtm(M).dtype == np.float64
```

Comment:
From gh-18250:
> Yes, when the input is real floating, the complex output will be twice the itemsize of the input, but that should not be the case when the input is complex. The real part of the input needs to be the same as the real part of the output.

This test and the other `float128`/`complex256` tests below are not aligned with that rule. Was there a discussion of exceptions? I understand that LAPACK does not work with these precisions, but I thought we should still preserve the input type. For instance:
```python3
import numpy as np
from scipy import linalg
A = np.eye(3, dtype=np.float128)
linalg.expm(A).dtype  # float128
```

Anyway, these tests are now failing because the function preserves the input precision rather than halving it. Please confirm that the tests should be adjusted (or left as they were, rather).

Reply:
According to https://github.com/scipy/scipy/pull/18385#issuecomment-2329725333, the tests were correct. Implementation adjusted accordingly (even simpler!).